### PR TITLE
1828: Implement D&H teleport ability

### PR DESCRIPTION
**Pull request by crericha**

### CODE REVIEW

File: lib/engine/ability/README.md [hunk 1]
```diff
@@ -130,6 +130,8 @@ Lay a tile and place a station token without connectivity
   teleport destination.
 - `tiles`: An array of tile numbers which may be placed at the
   teleport destination.
+- `cost`: Cost to use the telport ability.
```
Comment (michaeljb) [inline lib/engine/ability/README.md:133]: typo

File: lib/engine/step/tracker.rb [hunk 2]
```diff
@@ -91,9 +91,20 @@ def lay_tile(action, extra_cost: 0, entity: nil, spender: nil)
...
+          next if !ability.hexes.include?(hex.id) || !ability.tiles.include?(tile.name)
+
+          teleport = true
+          free = true if ability.free_tile_lay
+          if ability.cost
+            spender.spend(ability.cost, @game.bank) if ability.cost.positive?
+            @log << "#{spender.name} spends #{@game.format_currency(ability.cost)} and teleports to #{hex.name}"
+          end
+
+          ability.use!
```
Comment (michaeljb) [inline lib/engine/step/tracker.rb:103]: `process_lay_tile` in `special_track.rb` already calls `ability.use!`. Would it be possible for `lay_tile` to take teleport and cost as parameters, and get rid of the `.abilities(:teleport)` call here?

https://github.com/crericha/18xx/blob/fbc83f1fa66e894c5c883bc643ced201dbaf0e49/lib/engine/step/special_track.rb#L23-L28

File: lib/engine/ability/README.md (hunk 1 above)
Comment (crericha) [reply lib/engine/ability/README.md]: Fixed.

File: lib/engine/step/tracker.rb (hunk 2 above)
Comment (crericha) [reply lib/engine/step/tracker.rb:103]: I didn't add ability.use!. It was already there. (line 94)

There are other games that allow teleport without a tile lay (1849 comes to mind), so I think the teleport ability is still needed?

Comment (michaeljb) [reply lib/engine/step/tracker.rb:103]: > I didn't add ability.use!. It was already there (line 94). Is it something that should be removed?

Oh sorry, I missed that! Yeah looks like we need to leave it

File: lib/engine/ability/README.md [hunk 3]
```diff
@@ -130,6 +130,8 @@ Lay a tile and place a station token without connectivity
   teleport destination.
 - `tiles`: An array of tile numbers which may be placed at the
   teleport destination.
+- `cost`: Cost to use the teleport ability.
+- `fee_tile_lay`: If true, the tile is laid with 0 cost. Default false.
```
Comment (tobymao) [inline lib/engine/ability/README.md:134]: free?

File: lib/engine/ability/teleport.rb [hunk 4]
```diff
@@ -5,11 +5,13 @@
...
     class Teleport < Base
-      attr_reader :hexes, :tiles
+      attr_reader :hexes, :tiles, :cost, :free_tile_lay
 
-      def setup(hexes:, tiles:)
+      def setup(hexes:, tiles:, cost: nil, free_tile_lay: false)
         @hexes = hexes
         @tiles = tiles
+        @cost = cost
+        @free_tile_lay = free_tile_lay
```
Comment (tobymao) [inline lib/engine/ability/teleport.rb:14]: safer to do free_tile_lay: nil

free_tile_lay = free_tile_lay || false

**Review by tobymao — APPROVED** [review]

**Pull request merged**